File: stdlib/lua/hull/tests/test_search.lua
-- test_search.lua — Tests for hull.search
--
-- Tests FTS5 full-text search wrapper.
-- Requires db global (run via C test harness with caps).

local search = require('hull.search')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

local function assert_true(v, msg)
    if not v then
        error((msg or "") .. " expected true, got " .. tostring(v))
    end
end

-- ── create_index and basic operations ───────────────────────────────

test("create_index: creates FTS5 table", function()
    search.create_index("articles", {"title", "body"})
    -- Should not error; table created
end)

test("index: insert document", function()
    search.index("articles", "1", {title = "Hello World", body = "This is a test article about searching"})
    search.index("articles", "2", {title = "Lua Programming", body = "Learn Lua programming language"})
    search.index("articles", "3", {title = "World News", body = "Breaking news from around the world"})
end)

test("query: basic text search", function()
    local results = search.query("articles", "world")
    assert_true(#results >= 2, "expected at least 2 results")
    -- Results should have id and rank
    assert_true(results[1].id ~= nil, "expected id field")
    assert_true(results[1].rank ~= nil, "expected rank field")
end)

test("query: specific term", function()
    local results = search.query("articles", "lua")
    assert_eq(#results, 1)
    assert_eq(results[1].id, "2")
end)

test("query: no results", function()
    local results = search.query("articles", "xyznonexistent")
    assert_eq(#results, 0)
end)

test("query: with limit", function()
    local results = search.query("articles", "world", { limit = 1 })
    assert_eq(#results, 1)
end)

-- ── remove ──────────────────────────────────────────────────────────

test("remove: deletes document", function()
    search.remove("articles", "3")
    local results = search.query("articles", "news")
    assert_eq(#results, 0)
end)

-- ── drop_index ──────────────────────────────────────────────────────

test("drop_index: removes table", function()
    search.drop_index("articles")
    -- Querying after drop should error
    local ok = pcall(search.query, "articles", "test")
    assert_eq(ok, false)
end)

-- ── validation ──────────────────────────────────────────────────────

test("create_index: rejects invalid name", function()
    local ok = pcall(search.create_index, "bad name!", {"col"})
    assert_eq(ok, false)
end)

test("create_index: rejects _hull_ prefix", function()
    local ok = pcall(search.create_index, "_hull_test", {"col"})
    assert_eq(ok, false)
end)

test("create_index: rejects invalid column names", function()
    local ok = pcall(search.create_index, "test", {"bad col!"})
    assert_eq(ok, false)
end)

-- ── snippet and highlight ───────────────────────────────────────────

test("query: with snippet", function()
    search.create_index("docs", {"title", "content"})
    search.index("docs", "1", {title = "Guide", content = "This is a comprehensive guide to full text searching in databases"})
    local results = search.query("docs", "guide", {
        snippet = { column = 2, tokens = 10, before = "<b>", after = "</b>" }
    })
    assert_true(#results >= 1, "expected at least 1 result")
    search.drop_index("docs")
end)

-- Return results for C test harness
return {pass = pass, fail = fail}

File: stdlib/js/hull/tests/test_search.js
// test_search.js - Tests for hull:search
//
// Tests FTS5 full-text search wrapper.
// Requires db (run via C test harness with caps).

import { search } from "hull:search";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

function assertTrue(v, msg) {
    if (!v)
        throw new Error((msg || "") + " expected truthy, got " + v);
}

// ── createIndex and basic operations ────────────────────────────────

test("createIndex: creates FTS5 table", () => {
    search.createIndex("articles", ["title", "body"]);
});

test("index: insert document", () => {
    search.index("articles", "1", { title: "Hello World", body: "This is a test article about searching" });
    search.index("articles", "2", { title: "Lua Programming", body: "Learn Lua programming language" });
    search.index("articles", "3", { title: "World News", body: "Breaking news from around the world" });
});

test("query: basic text search", () => {
    const results = search.query("articles", "world");
    assertTrue(results.length >= 2, "expected at least 2 results");
    assertTrue(results[0].id !== undefined, "expected id field");
    assertTrue(results[0].rank !== undefined, "expected rank field");
});

test("query: specific term", () => {
    const results = search.query("articles", "lua");
    assertEq(results.length, 1);
    assertEq(results[0].id, "2");
});

test("query: no results", () => {
    const results = search.query("articles", "xyznonexistent");
    assertEq(results.length, 0);
});

test("query: with limit", () => {
    const results = search.query("articles", "world", { limit: 1 });
    assertEq(results.length, 1);
});

// ── remove ──────────────────────────────────────────────────────────

test("remove: deletes document", () => {
    search.remove("articles", "3");
    const results = search.query("articles", "news");
    assertEq(results.length, 0);
});

// ── dropIndex ───────────────────────────────────────────────────────

test("dropIndex: removes table", () => {
    search.dropIndex("articles");
    let threw = false;
    try { search.query("articles", "test"); } catch (e) { threw = true; }
    assertTrue(threw, "expected error after drop");
});

// ── validation ──────────────────────────────────────────────────────

test("createIndex: rejects invalid name", () => {
    let threw = false;
    try { search.createIndex("bad name!", ["col"]); } catch (e) { threw = true; }
    assertTrue(threw, "expected error for invalid name");
});

test("createIndex: rejects _hull_ prefix", () => {
    let threw = false;
    try { search.createIndex("_hull_test", ["col"]); } catch (e) { threw = true; }
    assertTrue(threw, "expected error for _hull_ prefix");
});

test("createIndex: rejects invalid column names", () => {
    let threw = false;
    try { search.createIndex("test", ["bad col!"]); } catch (e) { threw = true; }
    assertTrue(threw, "expected error for invalid column");
});

// A SQL keyword passes IDENT_RE -- it is a plain identifier -- so nothing
// above catches it, and `reindex` interpolates the source table and column
// names UNPREFIXED. Without this branch the failure surfaces as a bare
// SQLite syntax error naming neither the caller nor the word. Case-
// insensitive, because SQL's grammar is.
test("createIndex: rejects a SQL keyword as a column name", () => {
    let threw = false;
    try { search.createIndex("posts", ["from"]); } catch (e) { threw = true; }
    assertTrue(threw, "expected error for keyword column name");
});

test("reindex: rejects a SQL keyword as the source table", () => {
    let threw = false;
    try {
        search.reindex("posts", "order", { columns: { title: "title" } });
    } catch (e) { threw = true; }
    assertTrue(threw, "expected error for keyword source table");
});

test("createIndex: keyword check is case-insensitive", () => {
    let threw = false;
    try { search.createIndex("Select", ["col"]); } catch (e) { threw = true; }
    assertTrue(threw, "expected error for lower/mixed-case keyword");
});

// ── snippet ─────────────────────────────────────────────────────────

test("query: with snippet", () => {
    search.createIndex("docs", ["title", "content"]);
    search.index("docs", "1", { title: "Guide", content: "This is a comprehensive guide to full text searching in databases" });
    const results = search.query("docs", "guide", {
        snippet: { column: 2, tokens: 10, before: "<b>", after: "</b>" }
    });
    assertTrue(results.length >= 1, "expected at least 1 result");
    search.dropIndex("docs");
});

export default { pass, fail };
